src/responder.ts

```diff
@@ -201,77 +204,67 @@ export class EthereumTransactionMiner {
+                reject(new BlockTimeoutError(`No new block received too long; provider unresponsive.`));
+            });
+        });
 
         try {
             // First, wait to get at least 1 confirmation, but throw an error if the transaction is stuck
             // (that is, new blocks are coming, but the transaction is not included)
-            await Promise.race([
-                firstConfirmationPromise,
-                firstConfirmationTimeoutPromise,
-                noNewBlockPromise
-            ]);
+            await Promise.race([firstConfirmationPromise, noNewBlockPromise]);
         } finally {
             // Make sure any pending CancellablePromise is released
             firstConfirmationPromise.cancel();
-            firstConfirmationTimeoutPromise.cancel();
-            noNewBlockPromise.cancel();
         }
     }
 
-
     /**
      * Resolves after the transaction `txHash` receives `confirmationsRequired`.
-     * Rejects with `NoNewBlockError` the provider does not receive a new block for `newBlockTimeout` milliseconds (starting from `timeLastBlockReceived`).
+     * Rejects with `BlockTimeoutError` if the `this.blockTimeoutDetector` emits a `BLOCK_TIMEOUT_EVENT`.
      * Rejects with `ReorgError` if the transaction is not found by the provider.
      */
-    public async waitForEnoughConfirmations(txHash: string, timeLastBlockReceived: number) {
+    public async waitForEnoughConfirmations(txHash: string) {
         // Promise that waits for enough confirmations before declaring success
-        const enoughConfirmationsPromise = waitForConfirmations(this.provider, txHash, this.confirmationsRequired, true);
+        const enoughConfirmationsPromise = this.confirmationObserver.waitForConfirmations(
+            txHash,
+            this.confirmationsRequired,
+            null,
+            true
+        );
 
         // ...but stop with error if no new blocks come for too long
-        const noNewBlockPromise = rejectIfAnyBlockTimesOut(
-            this.provider,
-            timeLastBlockReceived,
-            this.newBlockTimeout,
-            this.pollInterval
-        );
+        const noNewBlockPromise = new Promise((_, reject) => {
+            this.blockTimeoutDetector.once(BlockTimeoutDetector.BLOCK_TIMEOUT_EVENT, () => {
+                reject(new BlockTimeoutError(`No new block received too long; provider unresponsive.`));
+            });
+        });
 
         try {
             // Then, wait to get at enough confirmations; now only throw an error if there is a reorg
-            await Promise.race([
-                enoughConfirmationsPromise,
-                noNewBlockPromise
-            ]);
+            await Promise.race([enoughConfirmationsPromise, noNewBlockPromise]);
         } finally {
             // Make sure any pending CancellablePromise is released
             enoughConfirmationsPromise.cancel();
```

Comment:
Again, doesnt the noNewBlockPromise need to be cancelled?

Reply:
Solved in ffc3e167ad041682416a75ae73c07cb439831e5b